Cannot subcribe to specific podcast when storage is on FAT32 external SD card: open failed: EINVAL

I have six podcasts already subscribed without problems. When attempting to subscribe to [The Guardian's Brexit Means...](https://www.theguardian.com/politics/series/brexit-means/podcast.xml), I got a message in Antennapod's download log saying "IO Error: open failed: EINVAL (Invalid argument)". The message did not specify which file name it complained about.

The only bug looking vaguely relevant I found was #268, but that's supposed to be solved, and the podcast's name doesn't contain a colon. However, it does contain an apostrophe, which may or may not be legal in FAT32.

Sure enough, if I change Antennapod's storage to internal I can subscribe to the podcast.

**App version:** 1.6.2.3 (from F-Droid)

**Android version**: 6.0.1 ([Fairphone Open 17.07.6](https://code.fairphone.com/contentx/2017-07-27_fairphone-open-17.07_release.html), not rooted, without Google services)

**Device model**: [Fairphone 2](https://www.fairphone.com/en/)

**Expected behaviour**: The podcast should be subscribed, no matter the storage back-end Antennapod uses.

**Current behaviour**: If storage is a FAT32 external SD card, the podcast is not subscribed, and a message saying "IO Error: open failed: EINVAL (Invalid argument)" appears in the download log.

**First occured**: Version 1.6.2.3 was already installed the first time I tried, on August 5, 2017.

**Steps to reproduce**:

1. Go to Settings -> Choose Data Folder.
1. Select your external SD card, formatted with FAT32 (e.g. `/storage/1234-ABCD/`).
1. Go to Subscriptions -> Add Podcast.
1. Enter `https://www.theguardian.com/politics/series/brexit-means/podcast.xml` in "Add Podcast by URL"
4.1. Alternatively, click "Search FYYD", enter `guardian brexit` and click the first result.
1. Click "Subscribe".

**Environment**: Auto Download is enabled on WiFi. When Antennapod is configured to use the external FAT32 SD card, the download fails. When Antennapod is configured to use internal storage, it succeeds.

**Stacktrace/Logcat** (with timestamps and UID:s stripped to save you from scrolling to the right):
```
08-07 21:07:49.758 32371 15358 W System.err: java.io.IOException: open failed: EINVAL (Invalid argument)
W System.err:    at java.io.File.createNewFile(File.java:939)
W System.err:    at de.danoeh.antennapod.core.service.download.HttpDownloader.download(HttpDownloader.java:186)
W System.err:    at de.danoeh.antennapod.core.service.download.Downloader.call(Downloader.java:43)
W System.err:    at de.danoeh.antennapod.core.service.download.Downloader.call(Downloader.java:14)
W System.err:    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
W System.err:    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
W System.err:    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
W System.err:    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
W System.err:    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
W System.err:    at java.lang.Thread.run(Thread.java:818)
W System.err: Caused by: android.system.ErrnoException: open failed: EINVAL (Invalid argument)
W System.err:    at libcore.io.Posix.open(Native Method)
W System.err:    at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
W System.err:    at java.io.File.createNewFile(File.java:932)
W System.err:    ... 9 more
```

--- a/app/src/androidTest/java/de/test/antennapod/util/FilenameGeneratorTest.java
+++ b/app/src/androidTest/java/de/test/antennapod/util/FilenameGeneratorTest.java
@@ -34,6 +34,16 @@ public class FilenameGeneratorTest extends AndroidTestCase {
 		createFiles(result);
 	}
 
+	public void testFeedTitleContainsApostrophe() {
+		String result = FileNameGenerator.generateFileName("Feed's Title ...");
+		assertEquals("Feeds Title", result);
+	}
+
+	public void testFeedTitleContainsDash() {
+		String result = FileNameGenerator.generateFileName("Left - Right");
+		assertEquals("Left Right", result);
+	}
+
 	/**
 	 * Tests if files can be created.
 	 * 

--- a/core/src/main/java/de/danoeh/antennapod/core/storage/DownloadRequester.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/storage/DownloadRequester.java
@@ -345,9 +345,7 @@ public class DownloadRequester {
         // Try to generate the filename by the item title
         if (media.getItem() != null && media.getItem().getTitle() != null) {
             String title = media.getItem().getTitle();
-            // Delete reserved characters
-            titleBaseFilename = title.replaceAll("[^a-zA-Z0-9 ._()-]", "");
-            titleBaseFilename = titleBaseFilename.trim();
+            titleBaseFilename = FileNameGenerator.generateFileName(title);
         }
 
         String URLBaseFilename = URLUtil.guessFileName(media.getDownload_url(),

--- a/core/src/main/java/de/danoeh/antennapod/core/util/FileNameGenerator.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/util/FileNameGenerator.java
@@ -1,18 +1,21 @@
 package de.danoeh.antennapod.core.util;
 
+import org.apache.commons.lang3.ArrayUtils;
+
 import java.util.Arrays;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /** Generates valid filenames for a given string. */
 public class FileNameGenerator {
 	
-	private static final char[] ILLEGAL_CHARACTERS = { '/', '\\', '?', '%',
-			'*', ':', '|', '"', '<', '>', '\n' };
-	static {
-		Arrays.sort(ILLEGAL_CHARACTERS);
-	}
+	private static final char[] validChars = (
+			"abcdefghijklmnopqrstuvwxyz" +
+			"ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+			"0123456789" +
+			" _-").toCharArray();
 	
 	private FileNameGenerator() {
-
 	}
 
 	/**
@@ -20,17 +23,17 @@ public class FileNameGenerator {
 	 * characters of the given string.
 	 */
 	public static String generateFileName(String string) {
-		StringBuilder builder = new StringBuilder();
+		StringBuilder buf = new StringBuilder();
 		for (int i = 0; i < string.length(); i++) {
 			char c = string.charAt(i);
-			if (Arrays.binarySearch(ILLEGAL_CHARACTERS, c) < 0) {
-				builder.append(c);
+			if(Character.isSpaceChar(c) && (buf.length() == 0 || Character.isSpaceChar(buf.charAt(buf.length()-1)))) {
+				continue;
+			}
+			if (ArrayUtils.contains(validChars, c)) {
+				buf.append(c);
 			}
 		}
-		return builder.toString().replaceFirst("  *$","");
+		return buf.toString().trim();
 	}
 
-	public static long generateLong(final String str) {
-		return str.hashCode();
-	}
 }

